hw1/hw1.py

```diff
@@ -0,0 +1,257 @@
+    X = (X - X_mean) / (X.max(axis=0) - X.min(axis=0))
+
+    y_mean = y.mean()
+    y = (y - y_mean) / (y.max() - y.min())
+    return X, y
+
+
+def apply_bias_trick(X):
+    """
+    Applies the bias trick to the input data.
+
+    Input:
+    - X: Input data (m instances over n features).
+
+    Returns:
+    - X: Input data with an additional column of ones in the
+        zeroth position (m instances over n+1 features).
+    """
+    m = X.shape[0]
+    ones = np.ones(m)
+    X = np.column_stack((ones, X))
+    return X
+
+
+def compute_cost(X, y, theta):
+    """
+    Computes the average squared difference between an observation's actual and
+    predicted values for linear regression.
+
+    Input:
+    - X: Input data (m instances over n features).
+    - y: True labels (m instances).
+    - theta: the parameters (weights) of the model being learned.
+
+    Returns:
+    - J: the cost associated with the current set of parameters (single number).
+    """
+
+    m = len(y)
+    predictions = X.dot(theta)
+    errors = np.subtract(predictions, y)
+    sqrErrors = np.square(errors)
+    J = 1 / (2 * m) * np.sum(sqrErrors)
+
+    return J
+
+
+def gradient_descent(X, y, theta, alpha, num_iters):
+    """
+    Learn the parameters of the model using gradient descent using
+    the training set. Gradient descent is an optimization algorithm
+    used to minimize some (loss) function by iteratively moving in
+    the direction of steepest descent as defined by the negative of
+    the gradient. We use gradient descent to update the parameters
+    (weights) of our model.
+
+    Input:
+    - X: Input data (m instances over n features).
+    - y: True labels (m instances).
+    - theta: The parameters (weights) of the model being learned.
+    - alpha: The learning rate of your model.
+    - num_iters: The number of updates performed.
+
+    Returns:
+    - theta: The learned parameters of your model.
+    - J_history: the loss value for every iteration.
+    """
+
+    theta = theta.copy()  # optional: theta outside the function will not change
+    m = len(y)
+    J_history = np.zeros(num_iters)
+
+    for iter in range(num_iters):
+        predictions = X.dot(theta)
+        errors = np.subtract(predictions, y)
+        sum_delta = (alpha / m) * X.transpose().dot(errors)
+        theta = theta - sum_delta
+        J_history[iter] = compute_cost(X, y, theta)
+
+    return theta, J_history
+
+
+def compute_pinv(X, y):
+    """
+    Compute the optimal values of the parameters using the pseudoinverse
+    approach as you saw in class using the training set.
+
+    #########################################
+    #### Note: DO NOT USE np.linalg.pinv ####
+    #########################################
+
+    Input:
+    - X: Input data (m instances over n features).
+    - y: True labels (m instances).
+
+    Returns:
+    - pinv_theta: The optimal parameters of your model.
+    """
+
+    pinv_theta = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)
```

Comment:
I guess we don't have to worry about `(X^T)X` being singular.

Reply:
Correct, this is actually the simplified version of `pinv`. The actual `pinv` uses the SVD-based algorithm.